src/backend/graphql/queries.js

```diff
@@ -0,0 +1,411 @@
+    };
+  },
+
+  literal(value: number | string | boolean | null): LiteralValue {
+    return {
+      type: "LITERAL",
+      data: value,
+    };
+  },
+
+  enumLiteral(value: string): EnumValue {
+    return {
+      type: "ENUM",
+      data: value,
+    };
+  },
+
+  list(data: Value[]): ListValue {
+    return {
+      type: "LIST",
+      data: data,
+    };
+  },
+
+  object(data: {[string]: Value}): ObjectValue {
+    return {
+      type: "OBJECT",
+      data: data,
+    };
+  },
+};
+
+/**
+ * An internal interface describing a formatting strategy for a sequence
+ * of GraphQL language tokens.
+ */
+interface LineFormatter {
+  // Format a line. For instance, a multi-line formatter might add some
+  // leading indentation.
+  format(line: string): string;
+
+  // Join multiple lines. For instance, a multi-line formatter might
+  // intersperse newlines, where a single-line formatter might
+  // intersperse spaces.
+  join(lines: string[]): string;
+
+  // Get a line formatter for the next level of nesting. For instance,
+  // if this object is a multi-line formatter, then this method might
+  // produce a line formatter with one extra level of indentation.
+  next(): LineFormatter;
+}
+
+/**
+ * Create a multi-line formatter that uses the given tab string (such as
+ * "\t" or "    ") for indentation.
+ */
+export function multilineFormatter(tab: string): LineFormatter {
+  function formatter(indentLevel: number) {
+    return {
+      format(line: string) {
+        const indentation = Array(indentLevel)
+          .fill(tab)
+          .join("");
+        return indentation + line;
+      },
+      next() {
+        return formatter(indentLevel + 1);
+      },
+      join(xs) {
+        return xs.join("\n");
+      },
+    };
+  }
+  return formatter(0);
+}
+
+/**
+ * Create a formatter that emits all tokens in one line, using spaces
+ * for separation.
+ */
+export function inlineFormatter(): LineFormatter {
+  const result = {
+    format: (line) => line,
+    next: () => result,
+    join: (xs) => xs.join(" "),
+  };
+  return result;
+}
+
+/*
+ * Map a stringification function across a list, and join the results
+ * with a formatter.
+ */
+function formatList<T>(
+  values: T[],
+  subformatter: (value: T, fmt: LineFormatter) => string,
+  fmt: LineFormatter
+): string {
+  return fmt.join(values.map((x) => subformatter(x, fmt)));
+}
+
+/*
+ * Map a stringification function across the values of an object, and
+ * join the keys and their corresponding results with a formatter.
+ */
+function formatObject<T>(
+  object: {[string]: T},
+  subformatter: (value: T, fmt: LineFormatter) => string,
+  fmt: LineFormatter
+): string {
+  function formatKey(k: string, fmt: LineFormatter): string {
+    return fmt.join([fmt.format(`${k}:`), subformatter(object[k], fmt.next())]);
+  }
+  return formatList(Object.keys(object), formatKey, fmt);
+}
+
+export const stringify = {
+  body(body: Body, fmt: LineFormatter): string {
```

Comment:
There's inconsistent usage of "stringify" vs "format" across this file. Maybe just standardize on "format", where "format" means "turn this thing into a string, using a provided LineFormatter". Then this const can be called "formats".

Possibly it would be relevant to define the type `Format<T>: function(T, LineFormatter) => string`.
Then `formats` is literally a collection of `Format` implementations

Reply:
Alternatively, I think using "stringify" as the consistent verb and "stringifiers" as the object containing many stringify functions may be even clearer, as it removes ambiguity with the two LineFormatters.